translator/metrics/oc_to_otlp.go

```diff
@@ -0,0 +1,390 @@
+	if len(md.Metrics) == 0 {
+		return resourceMetricsList
+	}
+
+	// TODO: Add InstrumentationLibrary field support
+	ilm := &otlpmetrics.InstrumentationLibraryMetrics{}
+	resourceMetrics.InstrumentationLibraryMetrics = []*otlpmetrics.InstrumentationLibraryMetrics{ilm}
+
+	ilm.Metrics = make([]*otlpmetrics.Metric, 0, len(md.Metrics))
+
+	for _, ocMetric := range md.Metrics {
+		if ocMetric == nil {
+			// Skip nil metrics.
+			continue
+		}
+
+		otlpMetric := metricToOtlp(ocMetric)
+
+		if ocMetric.Resource != nil {
+			// Add a separate ResourceMetrics item just for this metric since it
+			// has a different Resource.
+			separateRM := &otlpmetrics.ResourceMetrics{
+				Resource: translatorcommon.OCNodeResourceToOtlp(md.Node, ocMetric.Resource),
+				InstrumentationLibraryMetrics: []*otlpmetrics.InstrumentationLibraryMetrics{
+					{
+						Metrics: []*otlpmetrics.Metric{otlpMetric},
+					},
+				},
+			}
+			resourceMetricsList = append(resourceMetricsList, separateRM)
+		} else {
+			// Otherwise add the metrics to the first ResourceMetrics item.
+			ilm.Metrics = append(ilm.Metrics, otlpMetric)
+		}
+	}
+	return resourceMetricsList
+}
+
+func metricToOtlp(ocMetric *ocmetrics.Metric) *otlpmetrics.Metric {
+	metricDescriptor := descriptorToOtlp(ocMetric.GetMetricDescriptor())
+	otlpMetric := &otlpmetrics.Metric{
+		MetricDescriptor: metricDescriptor,
+	}
+	if metricDescriptor == nil {
+		return otlpMetric
+	}
+	commonLabels, labels := getOtlpLabels(ocMetric)
+	if commonLabels != nil {
+		otlpMetric.MetricDescriptor.Labels = commonLabels
+	}
+	setDataPoints(otlpMetric, ocMetric, labels)
+	return otlpMetric
+}
+
+// setDataPoints sets one of oltp metric datapoint slices based on metric type
+func setDataPoints(
+	otlpMetric *otlpmetrics.Metric,
+	ocMetric *ocmetrics.Metric,
+	labels [][]*otlpcommon.StringKeyValue,
+) {
+	switch ocMetric.MetricDescriptor.GetType() {
+	case ocmetrics.MetricDescriptor_GAUGE_INT64, ocmetrics.MetricDescriptor_CUMULATIVE_INT64:
+		otlpMetric.Int64DataPoints = getInt64DataPoints(ocMetric, labels)
+	case ocmetrics.MetricDescriptor_GAUGE_DOUBLE, ocmetrics.MetricDescriptor_CUMULATIVE_DOUBLE:
+		otlpMetric.DoubleDataPoints = getDoubleDataPoints(ocMetric, labels)
+	case ocmetrics.MetricDescriptor_GAUGE_DISTRIBUTION, ocmetrics.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		otlpMetric.HistogramDataPoints = getHistogramDataPoints(ocMetric, labels)
+	case ocmetrics.MetricDescriptor_SUMMARY:
+		otlpMetric.SummaryDataPoints = getSummaryDataPoints(ocMetric, labels)
+	default:
+	}
+}
+
+// getOtlpLabels scans OC metric labels and returns labels in OTLP format as two collections of StringKeyValue.
+// The first collection is a slice of labels that were found in all of OC metric timeseries.
+// The second collection is map of OTLP labels by OC timeseries indexes.
+func getOtlpLabels(ocMetric *ocmetrics.Metric) ([]*otlpcommon.StringKeyValue, [][]*otlpcommon.StringKeyValue) {
```

Comment:
Do we actually need this complex function? What if we do not try to find common labels and do not use the common labels field in OTLP? Presumably that would result in simpler code and unclear if performance will be worse.

Reply:
I was thinking about more about saving payload then performance. Yes, we can remove it, assuming that this solution is temporary and will be replaced by new internal data structure eventually.

Reply:
removed the common labels in the second commit